### DISCUSSION TITLE
Code splitting single JS files based on which functions are used by which parent components or files.

### DISCUSSION THREAD

**Discussion by jchabotamica** [webpack/webpack#18006] [Q&A]
I have been looking into ways to deeper code split my js files.  We are up to date with the latest versions of webpack, babel, etc.

We already do code splitting, but it seems to just do it on a file by file basis.  What I'd like to see is for it to further code split by which functions are used (not like pure tree shaking though because we have that too 😄 ).

As a visual:

src/index.ts
&nbsp;&nbsp;-- references function 1 from file A
&nbsp;&nbsp;--lazy load component A 
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;-- references function 2 from file A

I'm looking for this to split file A so that function 1 gets bundled with the main bundle, and function 2 gets bundles with the separate chunk which includes component A.

Also not, these are you using import/export syntax:
```
    //fileA
    export const function1 = () => {...};
    
    //index.ts
    import {function1} from 'fileA';
```

**Comment by alexander-akait** [comment]
Sorry it is not possible, anyway you can create a custom plugin to implement this, or just place require function/classes in each file and split them based on modules

**Reply by jchabotamica** [reply]
Sorry - can you expand on the latter part of your proposed solution?

**Reply by alexander-akait** [reply]
You can keep the classes in its own files and the same for function and use https://webpack.js.org/plugins/split-chunks-plugin/#splititchunkscachegroupstest.

**Reply by jchabotamica** [reply]
So webpack will take care of some of the code splitting, but there is still a decent amount of responsibility on the developers to ensure they create files separately.  So if we have a "common" file which 90% of the functions are used post-login (we use lazy loading using React.lazy() at the route level), but one function from that common file is also used in your src/index.tsx, then the entire utilities function is going to get bundled with the main bundle/chunk?

With your suggested approach, it sounds like it would still keep that entire "common" file together just in its own chunk.  There's no way to break it apart.

Can you confirm @alexander-akait ?

**Reply by alexander-akait** [reply]
Yes, To be honest, I wouldn’t recommend doing this at all, webpack has quite clever and complex merging and splitting logic, and most likely you will only worsen the performance of your application

**Reply by jchabotamica** [reply]
Assuming you mean you wouldn't recommend splitting off the entire "common" file to its own chunk, which I wouldn't want to do either.  However, would you recommend this practice for the intention of proper code splitting?  If you are importing modules into src/index.tsx (or any other components used before the lazy loaded routes), then ensure that those modules are concise and do not include other exported functions for other functionality.  For example, this "file A" should be broken apart and the responsibility is on the develoepr to do this - not webpack:
<img width="399" alt="image" src="https://github.com/webpack/webpack/assets/52013619/83640fe8-1d05-4613-8998-a7f8cee1cff2">

Or should this refactoring not be necessary if we're trying to truly take advantage of webpack's code splitting?  I'm trying to ensure that our developers are going down the right path, and also learning to do things the right way.  No matter how much I research, its very difficult to find training material around this specific topic.
